Sources/Core/KeystoreManager/BIP32HDNode.swift

```diff
@@ -120,130 +116,139 @@ public class HDNode {
-                if newPK == BigUInt(0) {
-                    if trueIndex < UInt32.max {
-                        return self.derive(index: index+1, derivePrivateKey: derivePrivateKey, hardened: hardened)
-                    }
-                    return nil
-                }
-                guard let privKeyCandidate = newPK.serialize().setLengthLeft(32) else {return nil}
-                guard SECP256K1.verifyPrivateKey(privateKey: privKeyCandidate) else {return nil }
-                guard let pubKeyCandidate = SECP256K1.privateToPublic(privateKey: privKeyCandidate, compressed: true) else {return nil}
-                guard pubKeyCandidate.bytes[0] == 0x02 || pubKeyCandidate.bytes[0] == 0x03 else {return nil}
-                guard self.depth < UInt8.max else {return nil}
-                let newNode = HDNode()
-                newNode.chaincode = cc
-                newNode.depth = self.depth + 1
-                newNode.publicKey = pubKeyCandidate
-                newNode.privateKey = privKeyCandidate
-                newNode.childNumber = trueIndex
-                guard let fprint = try? RIPEMD160.hash(message: self.publicKey.sha256())[0..<4] else {
-                    return nil
-                }
-                newNode.parentFingerprint = fprint
-                var newPath = String()
-                if newNode.isHardened {
-                    newPath = self.path! + "/"
-                    newPath += String(newNode.index % HDNode.hardenedIndexPrefix) + "'"
-                } else {
-                    newPath = self.path! + "/" + String(newNode.index)
-                }
-                newNode.path = newPath
-                return newNode
-            } else {
-                return nil // derive private key when is itself extended public key (impossible)
-            }
+            return deriveWithPrivateKey(index: index, hardened: hardened)
         } else { // deriving only the public key
-            var entropy: [UInt8] // derive public key when is itself public key
-            if index >= (UInt32(1) << 31) || hardened {
-                return nil // no derivation of hardened public key from extended public key
-            } else {
-                let hmac: Authenticator = HMAC(key: self.chaincode.bytes, variant: .sha2(.sha512))
-                var inputForHMAC = Data()
-                inputForHMAC.append(self.publicKey)
-                inputForHMAC.append(index.serialize32())
-                guard let ent = try? hmac.authenticate(inputForHMAC.bytes) else {return nil }
-                guard ent.count == 64 else { return nil }
-                entropy = ent
+            return deriveWithoutPrivateKey(index: index, hardened: hardened)
+        }
+    }
+
+    public func deriveWithoutPrivateKey(index: UInt32, hardened: Bool = false) -> HDNode? {
+        var entropy: [UInt8] // derive public key when is itself public key
+        if index >= (UInt32(1) << 31) || hardened {
```

Comment:
The `(UInt32(1) << 31)` is used 7 times across this file.
Let's move it out as a constant. Not a huge gain but at least we can assign a name to it like `maxIterationIndex` which is good.